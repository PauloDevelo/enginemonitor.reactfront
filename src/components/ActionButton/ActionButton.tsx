import React, { Fragment } from "react";
import { Button, Spinner } from 'reactstrap';
import { FormattedMessage } from 'react-intl';

import PropTypes from 'prop-types';

type Props = {
    color: string,
    isActing: boolean,
    message: FormattedMessage.MessageDescriptor,
    action?: () => void,
    className?: string,
    type?: "submit" | "button" | "reset",
    form?: string
};

const ActionButton = ({color, isActing, message, action, className, type, ...props}:Props) => {
    if (type === undefined){
        type = 'button';
    }

    return <Button type={type} color={color} onClick={action} className={className} disabled={isActing} {...props}>
        <FormattedMessage {...message} />
        {isActing && <Fragment>{' '}<Spinner size="sm" color="secondary" /></Fragment>}
    </Button>;
};

export default ActionButton;