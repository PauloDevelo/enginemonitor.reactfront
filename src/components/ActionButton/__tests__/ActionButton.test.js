import ignoredMessages from '../../../testHelpers/MockConsole';
import React from 'react';
import {  mount } from 'enzyme';

import ActionButton from '../ActionButton'

ignoredMessages.push("[React Intl] Could not find required `intl` object. <IntlProvider> needs to exist in the component ancestry. Using default message as fallback.");

describe('Component ActionButton', () =>{
    it('should render as a simple button', () => {
        const message = { id:"message", defaultMessage: "defaultMessage", description: "description"};
        const wrapper = mount(<ActionButton message={message}/>);

        expect(wrapper.text()).toEqual(message.defaultMessage);
        expect(wrapper.find('button').hasClass('btn')).toEqual(true);
    });

    it('should render as a simple button with the color color', () => {
        const message = { id:"message", defaultMessage: "defaultMessage", description: "description"};
        const wrapper = mount(<ActionButton message={message} color={"danger"}/>);

        expect(wrapper.text()).toEqual(message.defaultMessage);
        expect(wrapper.find('button').hasClass('btn-danger')).toEqual(true);
    });

    it('should render as a simple button disable with the danger color and a spinner', () => {
        const message = { id:"message", defaultMessage: "defaultMessage", description: "description"};
        const wrapper = mount(<ActionButton message={message} color={"danger"} isActing={true}/>);

        expect(wrapper).toMatchSnapshot();
    });

    it('should render as a simple button enable with the danger color and without a spinner', () => {
        const message = { id:"message", defaultMessage: "defaultMessage", description: "description"};
        const wrapper = mount(<ActionButton message={message} color={"danger"} isActing={false}/>);

        expect(wrapper).toMatchSnapshot();
    });

    it('should render a button clickable', () => {
        const message = { id:"message", defaultMessage: "defaultMessage", description: "description"};
        const action = jest.fn();
        const wrapper = mount(<ActionButton message={message} color={"danger"} isActing={false} action= {action}/>);

        wrapper.find('button').simulate('click');

        expect(action.mock.calls.length).toEqual(1);
    });
});