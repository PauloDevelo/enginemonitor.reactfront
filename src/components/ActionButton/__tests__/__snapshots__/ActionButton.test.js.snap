// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component ActionButton should render as a simple button disable with the danger color and a spinner 1`] = `
<IntlProvider
  defaultFormats={Object {}}
  defaultLocale="en"
  formats={Object {}}
  locale="en-US"
  messages={Object {}}
  onError={[Function]}
  textComponent={Symbol(react.fragment)}
  timeZone="Asia/Kuala_Lumpur"
>
  <ActionButton
    color="danger"
    isActing={true}
    message={
      Object {
        "defaultMessage": "defaultMessage",
        "description": "description",
        "id": "message",
      }
    }
  >
    <Button
      color="danger"
      disabled={true}
      tag="button"
      type="button"
    >
      <button
        aria-label={null}
        className="btn btn-danger disabled"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        <FormattedMessage
          defaultMessage="defaultMessage"
          description="description"
          id="message"
        >
          defaultMessage
        </FormattedMessage>
         
        <Spinner
          color="secondary"
          size="sm"
          tag="div"
          type="border"
        >
          <div
            className="spinner-border-sm spinner-border text-secondary"
            role="status"
          >
            <span
              className="sr-only"
            >
              Loading...
            </span>
          </div>
        </Spinner>
      </button>
    </Button>
  </ActionButton>
</IntlProvider>
`;

exports[`Component ActionButton should render as a simple button enable with the danger color and without a spinner 1`] = `
<IntlProvider
  defaultFormats={Object {}}
  defaultLocale="en"
  formats={Object {}}
  locale="en-US"
  messages={Object {}}
  onError={[Function]}
  textComponent={Symbol(react.fragment)}
  timeZone="Asia/Kuala_Lumpur"
>
  <ActionButton
    color="danger"
    isActing={false}
    message={
      Object {
        "defaultMessage": "defaultMessage",
        "description": "description",
        "id": "message",
      }
    }
  >
    <Button
      color="danger"
      tag="button"
      type="button"
    >
      <button
        aria-label={null}
        className="btn btn-danger"
        onClick={[Function]}
        type="button"
      >
        <FormattedMessage
          defaultMessage="defaultMessage"
          description="description"
          id="message"
        >
          defaultMessage
        </FormattedMessage>
      </button>
    </Button>
  </ActionButton>
</IntlProvider>
`;
